### PR TITLE
Merge some of the commits from kde-master

### PR DESCRIPTION
**Pull request by akreuzkamp**
This review request includes 16 of 27(resp. 21 applicable) commits that were done to kde-master since @Plaristote did his fork.

### CODE REVIEW

**Comment by Plaristote** [comment]
I see we worked on the same thing this morning. Luckily we only merged different commits, except for a3e8bf1.
I'll stop working on this branch for now, you can check #29 to see which commits I've already worked on.

**Comment by ChALkeR** [comment]
This introduces an LGPL-licensed file into the bundle,  without mentioning it in the actual LICENSE. This should be fixed in some way.

**Comment by ChALkeR** [comment]
Another note — can we have meaningful commit messages, please? =)

**Comment by ChALkeR** [comment]
@akreuzkamp Could you rebase this, please, and make the commits a bit more clear (i.e. without the prefix).

**Comment by ChALkeR** [comment]
02822c9 — Merged in #29.
3da78c7 — Merged in #29.
785965c — Merged in #29.
3fab595 —
1c12ea4 — Merged in #29.
d35f213 — Merged in #29.
c3f1f39 — Merged in #29.
12b725e — Merged in #29.
4c9e432 — Merged in #29.
48db468 — Merged in #29.
6022424 —
79a05ff —
2a7202f —
7dad8e8 —
f401c53 —
44bea45 —

**Comment by ChALkeR** [comment]
I'm rebasing this and excluding the already merged commits.

**Comment by ChALkeR** [comment]
The tests fail with:
```
TypeError: component.createObject is not a function
    at loadQMLTree (QMLEngine.js:120)
    at loadQML (QMLEngine.js:103)
    at loadQml (common.js:17)
```

**Comment by ChALkeR** [comment]
Looks to happen mainly because of `QMLBaseObject`.

File: src/qtcore/qml/elements/Component.js [hunk 1]
```diff
@@ -76,6 +76,7 @@ function QMLComponent(meta) {
         this.finalizeImports(this.$context);
     }
 }
+inherit(QMLComponent, QMLBaseObject);
```
Comment (ChALkeR) [inline src/qtcore/qml/elements/Component.js]: Most of the fails happen because of this change.

**Comment by ChALkeR** [comment]
The problem is that `inherit` overrides the prototype _after_ some methods have been set on the prototype.

**Comment by ChALkeR** [comment]
@akreuzkamp, please review my last commit here, it fixes things.

**Comment by ChALkeR** [comment]
The changes here are pretty straightforward.

First commit moves TextArea (from QtQuick.Controls) to QtQuick as TextEdit, and makes TextArea from QtQuick.Controls an alias for that (both are multi-line text input fields). This seems reasonable to me, it gives us a TextEdit implementation for now, which is important. This should be improved in the future, and differences between TextEdit and TextArea should be implemented, but this is out of scope for now.

Second commit implements prototype-based inheritance, which is also a good thing.

Third commit (the one by me) just fixes the second one so that manually assigned prototype methods that were called _before_ `registerQmlType` do not get lost.

There are still more work that needs to be done (but I would prefer to see that in separate PRs), e.g. `Button` is not described correctly, but it's not the fault of this commit. See #48 for details. Another one would be to re-check manual calls of constructor methods and (most likely) removal of most of those.

@Plaristote, @henrikrudstrom, could someone please give this a quick review? I want to land this on master.

Btw, it's the port of  6022424 that you were discussing here: https://github.com/qmlweb/qmlweb/issues/20#issuecomment-193214035.

**Comment by ChALkeR** [comment]
Hm. On the other hand, this doesn't include `getAttachedObject` from 6022424 for whatever reason, so the commit message should be changed. This is just the inheritance.

**Comment by Plaristote** [comment]
Everything looks good to me here, though I'm wondering about something in these kind of changes:
```javascript
registerQmlType({
  module:   'QtQuick',
  name:     'State',
  versions: /.*/,
  baseClass: QMLBaseObject,
  constructor: function QMLState(meta) {
    QMLBaseObject.call(this, meta);
```
The redundancy of QMLBaseObject is unsettling to me. I can live with it, I'm sure that trying to change this would go beyond the scope of this PR. But it sounds to me like everything done by `QMLBaseObject.call(this, meta)` should be done by checking the baseClass sent to registerQmlType.

**Comment by ChALkeR** [comment]
@Plaristote Yes, I was talking about those in https://github.com/qmlweb/qmlweb/pull/28#issuecomment-193765917:
> Another one would be to re-check manual calls of constructor methods and (most likely) removal of most of those.

I will file a follow-up PR to clean those things up a bit.

**Comment by ChALkeR** [comment]
@Plaristote I opened #133 so we don't forget about that =).

**Comment by ChALkeR** [comment]
Argh. Forgot to amend the commit message. Force-push time!

**Comment by akreuzkamp** [comment]
Ok, I think I missed quite some discussion here...
1. Missing getAttachedProperty() method: That got lost during the merge somehow. We need to readd that to the commit.
2. redundant constructor calls: I don't see a way to reduce that. `baseClass: QMLBaseObject` alters the prototype and thus makes some inherited methods available. So it alters the "class" QMLBaseObject. `QMLBaseObject.call(this, meta);` calls the base constructor. It alters the object and needs to be done everytime we create an object, not only once when we register the type. Those two lines do fundamentally different things. I don't see a way, removing one of them. Or does javascript provide a way to call base constructors automatically and I just don't know that?
3. @ChALkeR's fix: I'm not sure... What you did, seems like a workaround to me. Shouldn't we instead just fix the broken things? To me the reason for the failures is simple: inherit was called to late. It needs to be called before adding things to the prototype.

**Comment by Plaristote** [comment]
Wouldn't something like that be ok ?
```js
function registerQmlType(options) {
 // ...
  var constructor = function(meta) {
    options.baseClass.call(this, meta);
    options.constructor.call(this, meta);
  }
}
```

**Comment by ChALkeR** [comment]
@akreuzkamp 
> @ChALkeR's fix: I'm not sure... What you did, seems like a workaround to me. Shouldn't we instead just fix the broken things? To me the reason for the failures is simple: inherit was called to late. It needs to be called before adding things to the prototype.

I don't think that elements defining their methods on the prototype should be considered "broken", they were just not compatible with this new change. Atm, this change allows more flexibility with how the modules are defined without any significant drawbacks. We could reconsider that and refactor later, I don't think that we should refactor things before everything else is merged.

**Comment by akreuzkamp** [comment]
@Plaristote: Yes, ok, that works of course :)

@ChALkeR: I don't think that, either. ;)
The only thing, that is required, is to call inherit first. :) Your solution is a bit too "automagic" for my taste, because it moves a lot of stuff around. I'd prefer to just require inherit to be called first.
But you're right about the "merge things first." We can refactor that later. :)

**Pull request merged**